dbms/src/Storages/ReadInOrderOptimizer.cpp

```diff
@@ -0,0 +1,111 @@
+#include <Storages/ReadInOrderOptimizer.h>
+#include <Storages/MergeTree/MergeTreeData.h>
+#include <Functions/IFunction.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+
+ReadInOrderOptimizer::ReadInOrderOptimizer(
+    const ManyExpressionActions & elements_actions_,
+    const SortDescription & required_sort_description_,
+    const SyntaxAnalyzerResultPtr & syntax_result)
+    : elements_actions(elements_actions_)
+    , required_sort_description(required_sort_description_)
+{
+    if (elements_actions.size() != required_sort_description.size())
+        throw Exception("Sizes of sort description and actions are mismatched", ErrorCodes::LOGICAL_ERROR);
+
+    /// Do not analyze ARRAY JOIN result columns.
+    /// TODO: forbid more columns for analyzing.
```

Comment:
Why do we need to specifically forbid some columns? We won't have them in merge tree sorting key anyway. This needs a comment.

Reply:
Aliases to joined columns comes to order by as is and we can wrongly match them with order key of MergeTree.